### DISCUSSION TITLE
How to build the App layer with enabling CPX over UART2?

### DISCUSSION THREAD

**Discussion by Residualstress** [bitcraze/discussions#1248] [Q&A] [answered]
Hi, 
I'm trying to build the App layer with enabling CPX over UART2. Here is the detail:

### - What I have done: 
1. Completed this [tutorial](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/) to print the log variables of the Crazyflie on the Raspberry pi. The log variables are printed well. 
  But I'm not sure if the last error is normal or not.
     <img width="476" alt="image" src="https://github.com/bitcraze/discussions/assets/92587824/2aa5e8cb-ef42-4ea3-8d66-12d022c37ea7"><img width="488" alt="image" src="https://github.com/bitcraze/discussions/assets/92587824/58fce43a-6a66-430a-a230-a18f873fe17b">

2. To build the App layer with enabling CPX, I add the parameter to [crazyflie-firmware/examples/app_hello_world/app-config](https://github.com/bitcraze/crazyflie-firmware/blob/master/examples/app_hello_world/app-config) and then flashing it to CF.
 But I'm also not sure that's the right way to enable CPX over UART2.
     <img width="739" alt="image" src="https://github.com/bitcraze/discussions/assets/92587824/7df670c5-450b-418d-acc1-454ea9233837">
 And when I want to print the log variables, the script hangs on the "Connecting to UART". It's like I didn't set up UART correctly.
     ![WhatsApp 图像2024-03-25于16 16 45_c2a9716d](https://github.com/bitcraze/discussions/assets/92587824/2d7be18a-74b2-480d-888d-9e87a316d386)

### Questions:

1.  Is the last error in printing the log variables of the Crazyflie normal?
2. Is it correct to enable UART when build App layer by adding parameters to the [app-config](https://github.com/bitcraze/crazyflie-firmware/blob/master/examples/app_hello_world/app-config) file? Otherwise, what should I do to enable it.

Looking forward to your reply!

**Comment by knmcguire** [comment]
Could you also share which raspberry pi version you are using and which version of raspberry OS you have installed on the SD card? I've been trying to recreate your problem on the latest raspberry pi 5 (RaspOS 12) and I've hit a bug so we probably need to update the tutorial. Did you configure that with a baudrate of 115200 as well?

1. That is not exactly normla no, but I think it might be some timing issue.  Did you close the script with `ctrl-c`  ?The cflib doesn't have very good protection against closing of the script this way, and sometimes it will close everything off properly, but also a lot of times you will see this, where the connection is out but one part of the driver still tries to read a package. it's a bit dirty but in general not something to be overly concerned about I think
2. Yes that should be enough to enable all you need, although i think that only `DECK_CPX_HOST_ON_UART2` and `DECK_FORCE` should be necessary, but at least here you know for sure that CPX is enabled. Just mind that there is a bit of a copy/paste issue at the end of your app-config file but I'm not sure if that has anything to do with it?

**Reply by Residualstress** [reply, marked as answer]
Thanks for your reply.

> Could you also share which raspberry pi version you are using and which version of raspberry OS you have installed on the SD card? 

I'm using Raspberry Pi 4 with Raspbian GNU/Linux 10 (buster).

>Did you configure that with a baudrate of 115200 as well?

Yes, I set the baudrate of /dev/ttyS0 and Crazyflie UART2 in 115200. Because I notice the default baudrate of Crazyflie UART2 is 576000 which is not the common value. 

Meanwhile, to ensure that the correct baudrate is called when running `with SyncCrazyflie(URI) as scf:`, I also has to replace `self.cpx = CPX(UARTTransport(device, 576000))` with `self.cpx = CPX(UARTTransport(device, 115200))` in [cflib/crtp/serialdriver.py](https://github.com/bitcraze/crazyflie-lib-python/blob/master/cflib/crtp/serialdriver.py).

I think it would be nice to add an option for the baudrate when the user runs `with SyncCrazyflie(URI) as scf:` instead of manually changing the cflib.

> 1. That is not exactly normla no, but I think it might be some timing issue.  Did you close the script with `ctrl-c`  ?The cflib doesn't have very good protection against closing of the script this way, and sometimes it will close everything off properly, but also a lot of times you will see this, where the connection is out but one part of the driver still tries to read a package. it's a bit dirty but in general not something to be overly concerned about I think

I didn't close the script with 'ctrl-c'. The script was closed automatically. Actually, if I close it with 'ctrl-c', the error would not occur. 

BTW, every time I try to run the script again, it would gets stuck in `Connecting to UART on /dev/ttyS0 @ 115200 `. I had to reboot it to fix this.

> 2. Yes that should be enough to enable all you need, although i think that only `DECK_CPX_HOST_ON_UART2` and `DECK_FORCE` should be necessary, but at least here you know for sure that CPX is enabled. Just mind that there is a bit of a copy/paste issue at the end of your app-config file but I'm not sure if that has anything to do with it?

Thank you, I solved the copy/paste issue but the problem is still there.By checking the auto.conf in /crazyflie-firmware/examples/app_hello_world/build/include, I wanted to confirm whether both `DECK_CPX_HOST_ON_UART2` and `CONFIG_DECK_FORCE` are effective. The auto.conf showed `CONFIG_DECK_FORCE="none"` which means that `DECK_CPX_HOST_ON_UART2=cpxOverUART2` in [app-config](https://github.com/bitcraze/crazyflie-firmware/blob/master/examples/app_hello_world/app-config) is not working yet. So I tried to solve this problem by putting quotes around cpxOverUART2, i.e., `DECK_CPX_HOST_ON_UART2="cpxOverUART2"`. When I build and flash the app layer again, the log variables are printed smoothly.
![image](https://github.com/bitcraze/discussions/assets/92587824/9180edaa-015e-4075-a8e4-4d9faf98f578)

**Reply by knmcguire** [reply]
good find! I wasn't think of this but yes that is indeed necessary for string defines. 

Thanks for updating us on your fix!
